### PR TITLE
Primitive types and class literals

### PR DESCRIPTION
**Pull request by ashu-mehra**
This PR is an attempt to give java primitive types a "home".
It builds on that and creates class objects for primitive types,
and fixes few issues to make class literals for primitive types and their arrays work correctly.

### CODE REVIEW

File: integration-tests/src/it-in/snippets/ClassLiteralTests.java [hunk 1]
```diff
@@ -92,16 +97,12 @@ static void testMultiDimensionalArrayOfConcreteType() {
         } else {
             putchar('F');
         }
+        putchar('#');
     }
     public static void main(String args[]) {
-        /**
-         * Tests for class literals of primitive type are disabled as
```
Comment (ashu-mehra) [inline integration-tests/src/it-in/snippets/ClassLiteralTests.java:98]: Just for the record - this comment was incorrect. Before this PR these tests were failing and the real reason for failure was that we are not creating j.l.Class object for primitive types.

File: compiler/src/main/java/org/qbicc/type/PrimitiveType.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
+package org.qbicc.type;
+
+import java.util.function.Consumer;
+
+public class PrimitiveType {
```
Comment (dmlloyd) [inline compiler/src/main/java/org/qbicc/type/PrimitiveType.java:5]: The idea here I think is really good and cleans up a lot of ugly stuff.  A couple minor notes though:

- The fact that this class ends in `...Type` and resides in the `.type` package makes it appear as if this is part of the type system, but it is not.  Maybe just call it `Primitive`?
- Would this be better done as an `enum`?

File: compiler/src/main/java/org/qbicc/type/BooleanType.java [hunk 3]
```diff
@@ -50,4 +50,8 @@ public StringBuilder toString(final StringBuilder b) {
     public StringBuilder toFriendlyString(final StringBuilder b) {
         return b.append("bool");
     }
+
+    public PrimitiveType asJavaType() {
```
Comment (dmlloyd) [inline compiler/src/main/java/org/qbicc/type/BooleanType.java:54]: `asJavaType` is a bit at odds with the class name `PrimitiveType` IMO; they should be more consistent.

File: plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java [hunk 4]
```diff
@@ -33,17 +34,9 @@ public void accept(CompilationContext ctxt) {
...
-        // primitives
-        // [1] boolean.class
-        // [2] byte.class
-        // [3] short.class
-        // [4] char.class
-        // [5] int.class
-        // [6] float.class
-        // [7] long.class
-        // [8] double.class
-        // [9] void.class
```
Comment (dmlloyd) [inline plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: The class for primitives does not match this ordering, so it is very important to ensure that we don't have these values hard-coded anywhere. I think every such place would be marked with `TODO` or `XXX` though...

File: compiler/src/main/java/org/qbicc/type/PrimitiveType.java (hunk 2 above)
Comment (ashu-mehra) [reply compiler/src/main/java/org/qbicc/type/PrimitiveType.java:5]: Renamed `PrimitiveType` to `Primitive` and changed it to enum.

File: compiler/src/main/java/org/qbicc/type/BooleanType.java (hunk 3 above)
Comment (ashu-mehra) [reply compiler/src/main/java/org/qbicc/type/BooleanType.java:54]: Renamed `asJavaType` to `asPrimitive`.

File: plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java (hunk 4 above)
Comment (ashu-mehra) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: Sorry, I didn't get it. Can you please point out where is the mismatch in the ordering?

Comment (DanHeidinga) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: I think the the primitives do match but it's much more implicit now.  The primitives are added with `assignTypeIdToPrimitives()`, which does `PrimitiveType.forEach()` which in turn iterates the `PrimitiveType.TYPES` array using `for (PrimitiveType type : PrimitiveType.TYPES) {`.

As long as the order of elements in the `PrimitiveType.TYPES` doesn't change, this should match.  But it's super subtle.

Best to check everywhere using this and ensure it doesn't rely on the ordering.  (ie: that sub-int types typeid < int.typeid}

Comment (dmlloyd) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: Sorry, I thought you had `boolean` last, but maybe I just read it wrong without my glasses on.

Comment (ashu-mehra) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: still not clear - why are we worried about order of primitives anyway? is there relationship between typeid of primitives?

Are there any other places where we hard-code the typeid for primitives? Is that what we are are worried about - that the hard-coded values may not match with the order in `PrimitiveType`?

**Comment by ashu-mehra** [comment]
Updated commits as per the suggestions.

File: plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java (hunk 4 above)
Comment (dmlloyd) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: Yes, exactly; there may be places where type IDs for primitives were hard-coded so we should ensure that those all are replaced with references to `Primitive`.

Comment (ashu-mehra) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: ok, only place I could find was in `LLVMModuleNodeVisitor` where we lower the `TypeLiteral`. Do you know any other place that needs to be updated.

Comment (DanHeidinga) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: I don't see any other places - the ones that typically need to be changed are the ObjectModel helpers but so far they only care about where Object starts, where the interfaces start, and the arrays.

Comment (ashu-mehra) [reply plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/SupersDisplayBuilder.java:45]: @DanHeidinga thanks for confirming

File: compiler/src/main/java/org/qbicc/type/Primitive.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    CHAR("char"),
+    INT("int"),
+    FLOAT("float"),
+    LONG("long"),
+    DOUBLE("double"),
+    VOID("void");
+
+    private final String name;
+    private int typeId;
+    private ValueType type;
```
Comment (dmlloyd) [inline compiler/src/main/java/org/qbicc/type/Primitive.java:19]: Since these are set late and are shared variables, I think they should be `volatile`.

**Review by dmlloyd — APPROVED** [review]
I only have one more comment; once that's addressed I think this can be merged.

File: compiler/src/main/java/org/qbicc/type/Primitive.java (hunk 5 above)
Comment (ashu-mehra) [reply compiler/src/main/java/org/qbicc/type/Primitive.java]: Done

**Comment by ashu-mehra** [comment]
This is ready now.

**Review by DanHeidinga — APPROVED** [review]

**Pull request merged**